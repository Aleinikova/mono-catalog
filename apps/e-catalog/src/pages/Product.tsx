import { useRef } from 'react';
import { useSelector } from 'react-redux';

import {
  Button,
  makeStyles,
  shorthands,
  Spinner,
  Text,
  tokens,
  typographyStyles,
} from '@fluentui/react-components';

import { RootState } from '../store';
import { selectCategoryById } from '../store/categoriesStore';
import { useParams } from 'react-router-dom';
import {
  selectProductById,
  selectProductsStatus,
} from '../store/productsStore';
import { currencyFormatter } from '@mono-catalog/currency-formatter';
import { useGenerateFile } from '@mono-catalog/useGenerateFile';

const TEMPLATE_BASE64 =
  'VTJGc2RHVmtYMThxaWVSSWtpS29nRG5lWEFoWDl0dHFDdHVtSU40SXBTZWZPbkJVdTNzNlhFNlJyYnZUVU9UbjlDVjFpS0FRa3kzcjBUYzNTNmVGeG5DT2FHVjdHRi9KSk9KNThIYTFYeDNTb2czakpVM2EzM0pyVUZoSlFRY2dzZkNLRmhkZ0FSSGRscGlEVVNQQUpML2JxTTM1RHRPeWNERy9ML0dCM0dWTThpcDJ2Q1piS1Y0L091b29KeVdYT1JyUi8yeXQvSzYxdUxzRUtRY0VTS2hMUEIxeldMQ0xkb1N4dEVodmIrWU9lT0dST2hUTTlLdG00TG9aQnlxbzNYdVNERTdVL1RlbDFDL3g4VUhjSE00RWxMbVpDazNwNXJCTndvOXdGQlVkWkNkODNKaGVsU3ZLSWpkVGxOUjRjc05qSWUrOWFiTm43NDNvbWNLNWlLTCtQZEZOY1NsUG5GODIybFEvdWZXYmxUcVFLUzFZNXE0SmEzVHNHZzZEVEdKbkcvTUc1V0NsWUJoS3hnQ0Mrd2EvMGxLZ2gyWDBUK0paeGVLWjl4d0xncFYwRUVlTU5rRFlqRitJcEN4YnF4UlY3cmVwOHdvRng4RVpMMzNDZHFULzZHenk5UnR4dGVqWXdpRndGL0hJNi9JbTV6U2RLUnF5RWNDM2pNQklqVXdGZFlad1UrdnR6d1ovU2daWGxqR1B6MDUrOTM4WVBxcHBFWmkzUThtM2NMbjBQQmJzTkROV3FWKzZZN3oyd1VJTjZKcVNRM1BpdjhTa1NpM283VHh5dmp4Z2xGQWVCSk5oT0ZxSUxoeDNjdGVmdjF1aEJFaU1zK1dSNnJxMXlacUNLWjhVNkk5ZFFGVXNVSjNOTlBOakxXOU1QTzVvT24ySTAvRGRkYlRnYzBpcVFNOEJtY2M5T1l6SEFnODRKYkVQVzJUWGl6dVFqVFFJbEhkMHNNclY3UDV5Y0MyUG93TzRRSEhaQ0dxN0hsbmN6by9LK2h4UE1tMmZGZG1KaERpMHZ5VWlNbStPSkRIK1NGOVgxdlQ4NTR6ZlBKMlRGZERBSkUzNnBjS2d1Z2dWRWc3VndVNURoYlFCUlNPYytCeERlSlIzdDA0MXdwZXBmeUViZ014RzBDb1hIbklnMVJ3ZEQ0ZXJkRjJhQ0E4ZDlCZkpIL1AzYS8zSlBTYWxUNmVnVlptMTdLcExxMFVMZ3U4bFFkeCtVT0NEeFRpOURPbnJmRHZFTGszWkxoOUpqdUd2MWFVcWNDR0FlWTlWZkVmWUQ5MzIvcVdvZmhzM0oxL25nSWl0TGNSOXRiK2xGR0wxU1pDbEpMMD0';

const useStyles = makeStyles({
  root: {
    flexGrow: 1,
    display: 'grid',
    gridTemplateColumns: 'repeat(12, 1fr)',
    gridTemplateRows: '1fr',
    gridColumnGap: '32px',
    gridRowGap: '0px',
  },
  photo: {
    ...shorthands.gridArea(1, 1, 2, 6),
  },
  img: {
    width: '100%',
    objectFit: 'contain',
  },
  info: {
    display: 'flex',
    flexDirection: 'column',
    ...shorthands.gridArea(1, 1, 1, 13),
  },
  iframeContainer: {
    ...shorthands.gridArea(2, 1, 2, 13),
  },
  spinner: {
    height: '100%',
  },
  title: {
    display: 'block',
    marginBottom: tokens.spacingVerticalXXL,
    ...typographyStyles.largeTitle,
  },
  keyValueRow: {
    display: 'flex',
    marginBottom: tokens.spacingVerticalXXL,
  },
  key: {
    display: 'block',
    marginRight: tokens.spacingHorizontalXS,
    ...typographyStyles.body1Strong,
  },
  value: {
    display: 'block',
    ...typographyStyles.body1,
  },
  iframe: {
    height: '100%',
    width: '100%',
    minHeight: '600px',
  },
  content: {
    display: 'grid',
    gridTemplateColumns: 'repeat(12, 1fr)',
    gridTemplateRows: '2fr',
    gridColumnGap: '32px',
    gridRowGap: '16px',
    ...shorthands.gridArea(1, 6, 1, 13),
  },
  button: {
    marginBottom: tokens.spacingVerticalS,
  },
});

function Product() {
  const styles = useStyles();

  const { productId } = useParams();

  const product = useSelector((state: RootState) =>
    productId ? selectProductById(state, productId) : null
  );

  const status = useSelector(selectProductsStatus);

  const category = useSelector((state: RootState) =>
    product?.categoryId ? selectCategoryById(state, product?.categoryId) : null
  );

  const iframeContainer = useRef<HTMLIFrameElement | null>(null);

  const isLoading = status === 'loading' || status === 'idle';

  const {
    generateIframeLink,
    iframeLink,
    downloadFile,
    fileName,
    isGenerated,
  } = useGenerateFile(TEMPLATE_BASE64);

  const generateXML = () => {
    if (!product) {
      return;
    }

    const description = category?.fields?.map(
      (filed) => `${filed.title}: ${product[filed.name] || '-'}`
    );

    const xmlContent = `
    <Name>${product.name}</Name>
    <Price>${product.price} ${product.currency}</Price>
    <Inventory>${product.inventory}</Inventory>
    <Description>${description?.join(';  ') || ''}</Description>
    `;

    const finalXML = `<CustomXmlPart xmlns="http://schemas.officeatwork.com/2024/product">${xmlContent}</CustomXmlPart>`;

    generateIframeLink(finalXML);
  };

  if (isLoading) {
    return <Spinner />;
  }

  if (!product) {
    return <div>Not found</div>;
  }

  return (
    <div className={styles.root}>
      <div className={styles.photo}>
        <img
          className={styles.img}
          src={
            product?.imageUrl ||
            'https://upload.wikimedia.org/wikipedia/commons/thumb/6/65/No-Image-Placeholder.svg/1665px-No-Image-Placeholder.svg.png'
          }
          alt="product"
        />
      </div>
      <div className={styles.content}>
        <div className={styles.info}>
          <Text className={styles.title}>{product.name}</Text>
          <div className={styles.keyValueRow}>
            <Text className={styles.key}>Category:</Text>
            <Text className={styles.value}>{category?.name || 'Unknown'}</Text>
          </div>
          <div className={styles.keyValueRow}>
            <Text className={styles.key}>Inventory:</Text>
            <Text className={styles.value}>{product.inventory}</Text>
          </div>
          <div className={styles.keyValueRow}>
            <Text className={styles.key}>Price:</Text>
            <Text className={styles.value}>
              {currencyFormatter(+product.price, product.currency)}
            </Text>
          </div>
          {category?.fields &&
            category?.fields.map((field) => (
              <div className={styles.keyValueRow} key={field.name}>
                <Text className={styles.key}>{field.title}:</Text>
                <Text className={styles.value}>
                  {product[field.name] || '-'}
                </Text>
              </div>
            ))}
          <Button className={styles.button} onClick={generateXML}>
            Generate Document
          </Button>
          {isGenerated && (
            <Button
              appearance="primary"
              onClick={downloadFile}
            >{`Download ${fileName}`}</Button>
          )}
        </div>
        <div className={styles.iframeContainer} ref={iframeContainer}>
          {iframeLink && (
            <iframe
              className={styles.iframe}
              title="OfficeAtWorkTemplateChooser"
              src={iframeLink}
            />
          )}
        </div>
      </div>
    </div>
  );
}

export default Product;
