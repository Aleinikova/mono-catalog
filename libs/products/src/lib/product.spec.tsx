import { render } from '@testing-library/react';

import Product from './product';

const PRODUCT = {
  categoryId: 1,
  id: 1,
  name: 'iPhoneX',
  imageUrl:
    'https://www.freeiconspng.com/thumbs/iphone-x-pictures/apple-iphone-x-pictures-5.png',
  price: '1000',
  currency: 'usd',
  inventory: 10,
};

describe('Product', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<Product product={PRODUCT} />);
    expect(baseElement).toBeTruthy();
  });
});
