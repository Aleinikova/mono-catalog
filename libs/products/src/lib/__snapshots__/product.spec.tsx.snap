// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should match snapshot 1`] = `
<div
  className="fui-Card ___svmuv00_ds8hu60 f1p9o1ba f1sil6mw fifeqxg f899z7z f4h3tyx f18ur2pz f1lplnzb f10m5gbb fhftqfp f1k04kkk fxsr4vj fcvsdzp f22iagw f10pi13n f1ewtqcl f19n0e5 f1mdlcz9 frwkxtg f1n6gb5g f15yvnhg fo72kxq f13zj6fq f1nwj1ja f8rth92 flthirb f1lh990p ftkbnf5 f6czdpx f13hvwk3 f1i1u9k0 f1en4csx f1qnomq5 f2fl922 f1anhtl f1n2zcl3 f16v3d5c f1vx9l62 fobhde4 fx5r7kn f19chtn8 fuvs6re fy4glsf frsmuga fuldkky fxugw4r f1whvlc6 f16gxe2i fpgykix f1osi826 fzybk4o ftqa4ok f2hkw1w f8hki3x f1d2448m f1bjia2o ffh67wi f226i61 f13kzufm flujwa2 fsx75g8 f15bsgw9 f14e48fq f18yb2kv fd6o370 fh1cnn4 fy7oxxb fpukqih f184ne2d f99gebs f13b0oaq f1jvq617 f8t2bz6 f11unbnk fbd201q f12nqxso f1uguk4w f1dvezut fd0oaoj fjvm52t f1cwg4i8 f15fr7a0 fwsq40z f34ld9f fy0y4wt"
  data-testid="product-1"
  role="group"
>
  <button
    aria-haspopup="menu"
    className="fui-Button r1alrhcs ___aet0oh0_571hc70 fcgxt0o f17fgpbq fq4mcun f1cnd47f fhxju0i"
    data-tabster="{"restorer":{"type":1}}"
    id="menur0"
    onClick={[Function]}
    onContextMenu={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="1em"
      viewBox="0 0 20 20"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 6.5A1.75 1.75 0 1 1 10 3a1.75 1.75 0 0 1 0 3.5ZM10 17a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm-1.75-7a1.75 1.75 0 1 0 3.5 0 1.75 1.75 0 0 0-3.5 0Z"
        fill="currentColor"
      />
    </svg>
  </button>
  <div
    className="fui-CardPreview ___1hkdc12_193a1gh f10pi13n f1rzl30s f1chxnfo f17xb19f fecqxwr"
  >
    <img
      alt="Preview of a Word document: About Us - Overview"
      className="___1we4501_0000000 f1kle4es"
      src="https://www.freeiconspng.com/thumbs/iphone-x-pictures/apple-iphone-x-pictures-5.png"
    />
  </div>
  <div
    className="___1400c38_0000000 f22iagw f1869bpl fly5x3f"
  >
    <h2
      className="fui-Text ___c8i2lc0_1wzyky6 fk6fouc fod5ikn faaz57k fl43uef fpgzoln f1w7gpdv f6juhto f1mtd64y f1y7q3j9 f2jf649 fqcjy3b"
    >
      iPhoneX
    </h2>
    <span
      className="fui-Text ___nzdt000_ag2scu0 fk6fouc fkhj508 f1i3iumi figsok6 fpgzoln f1w7gpdv f6juhto f1mtd64y f1y7q3j9 f2jf649"
    >
      (
      10
      )
    </span>
  </div>
  <h2
    className="fui-Text ___r9uoz30_1njyo59 fk6fouc fod5ikn faaz57k fl43uef f17mccla ftgm304 f6juhto f1mtd64y f1y7q3j9 f2jf649"
  >
    $1,000.00
  </h2>
</div>
`;
